### PR TITLE
soc: arm: st_stm32: Cleanup gpio function code.

### PR DESCRIPTION
**Pull request by cybertale**
f0/f2/f3/f4/f7/l0/l4 series STM32 have mostly the same GPIO architecture
and can share the same code for GPIO manipulation. Only f1 series chips
have standalone architecture.
Functions of the external interrupt line control are also the same.
This patch extracts common code from them and put them into the 'common'
folder.

Functions of control GPIO of these series scattered in
soc/arm/st_stm32/stm32xx/ folders contain these functions:
stm32_gpio_flags_to_conf(), stm32_gpio_configure(), stm32_gpio_set(),
stm32_gpio_get, stm32_gpio_enable_int().
This patch move them into the soc_gpio-no-f1.c file and add them as
sources when these series chips are selected.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

### CODE REVIEW

**Comment by cybertale** [comment]
I tested this code on a stm32f407 based custom board.

**Review by erwango — COMMENTED** [review]
@cybertale, this was in my TODO list for a while, but never got the time to do it. This is a welcome change, thanks!

As a first feedback, can you look into including F1 in the change?
It will be easier if everything sits at the same place, and there is mainly 1 really divergent function (stm32_gpio_configure), this could be handled in the same file. stm32_gpio_flags_to_conf as similar structure and it could be tricked to be compatible.

Then, if you're ready for it, a nice addition would be to use STM32Cube LL API to even reduce the code base:
  - In function stm32_gpio_enable_int, use LL_SYSCFG_SetEXTISource/LL_GPIO_AF_SetEXTISource 
  - In stm32_gpio_set, use LL_GPIO_SetOutputPin/LL_GPIO_ResetOutputPin
  - In stm32_gpio_configure, LL_GPIO_SetPinMode, LL_GPIO_SetPinOutputType, ..
  - ...

**Comment by cybertale** [comment]
@erwango 
Hi erwango, 
Strange I did look into this thread yesterday but didn't see your reply, sorry for the late reply!
I'm very familiar with the F1 series chips, I'll see what I can do.

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+{
+	int direction = flags & GPIO_DIR_MASK;
+	int pud = flags & GPIO_PUD_MASK;
+
+	if (!pincfg) {
+		return -EINVAL;
+	}
+
+	if (direction == GPIO_DIR_OUT) {
+#ifdef CONFIG_SOC_SERIES_STM32F1X
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: It would be clearer to define a smart set of headers and remove all this ifdefery from this function
```suggestion
#ifdef CONFIG_SOC_SERIES_STM32F1X
#define PIN_MODE_OUTPUT STM32_MODE_OUTPUT | STM32_CNF_GP_OUTPUT | STM32_CNF_PUSH_PULL
#define PIN_MODE_INPUT ...
#define PIN_PUPD_PULLUP ...
...
#else
#define PIN_MODE_OUTPUT   STM32_MODER_OUTPUT_MODE
#define PIN_MODE_INPUT ...
#define PIN_PUPD_PULLUP ...
```

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+ * Open-RnD Sp, z o.o. (stm32f1/soc_gpio.c)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief
+ *
+ * Common manipulation functions for GPIO of
+ * stm32f0/f1/f2/f3/f4/f7/l0/l4 series.
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Update to "... GPIO of STM32 series"

**Review by erwango — CHANGES_REQUESTED** [review]
Few comments otherwise sounds good.
Adding LL would have been nice but can be done latter on, so is not blocking.

On which series have you been able to test this?

File: soc/arm/st_stm32/common/soc_gpio.c (hunk 1 above)
Comment (cybertale) [reply soc/arm/st_stm32/common/soc_gpio.c]: I just finished implementing these functions with LL library. 
I was just want to ask should I push only one commit using the LL library or two commits that the first one merge separate soc_gpio.c to one file, and the second commit change the implementation to use the LL library because one commit should do only one thing?

Comment (cybertale) [reply soc/arm/st_stm32/common/soc_gpio.c]: I have a f103zet6 based board and a f407zgt6 based board. Usually I test on both of them once I make some changes.

Comment (erwango) [reply soc/arm/st_stm32/common/soc_gpio.c]: > I just finished implementing these functions with LL library.
I was just want to ask should I push only one commit using the LL library or two commits that the first one merge separate soc_gpio.c to one file

Ok, great thanks. Please split into 2 commits, it will be easier to review.

Comment (erwango) [reply soc/arm/st_stm32/common/soc_gpio.c]: >  f103zet6 based board and a f407zgt6

Ok, thanks. I'll do some additional tests on the boards I own.

**Comment by cybertale** [comment]
Hi @erwango, I updated the code, how about this time?

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+#include <gpio.h>
+#include <gpio/gpio_stm32.h>
+#include <soc.h>
+#include "soc_syscfg_common.h"
+
+/* F1 series STM32 use AFIO register to configure interrupt generation
+ * and pinmux.
+ */
+#ifdef CONFIG_SOC_SERIES_STM32F1X
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Shouldn't this be removed in the LL commit?

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+ * Common manipulation functions for GPIO of stm32 series.
+ *
+ */
+
+#include <errno.h>
+
+#include <gpio.h>
+#include <gpio/gpio_stm32.h>
+#include <soc.h>
+#include "soc_syscfg_common.h"
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Is that still required in LL commit?

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+		} else {
+			LL_GPIO_SetAFPin_8_15(gpio, pin, altf);
+		}
+	}
+#endif /* CONFIG_SOC_SERIES_STM32F1X */
+
+	return 0;
+}
+
+int stm32_gpio_set(u32_t *base, int pin, int value)
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Could be moved in stm32 gpio driver (merged in gpio_stm32_write)

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+	} else {
+		LL_GPIO_ResetOutputPin(gpio, pin);
+	}
+
+	return 0;
+}
+
+int stm32_gpio_get(u32_t *base, int pin)
+{
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)base;
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Could be moved in stm32 gpio driver (merged in gpio_stm32_read)

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+	return (LL_GPIO_ReadInputPort(gpio) >> pin) & 0x1;
+}
+
+int stm32_gpio_enable_int(int port, int pin)
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Move to gpio driver. Could remain a separate function.

File: soc/arm/st_stm32/common/soc_syscfg_common.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2018 Song Qiang <songqiang1304521@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __SOC_SYSCFG_COMMON_H
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_syscfg_common.h]: Is that file still required?

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+{
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)base_addr;
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(altf);
+
+	pin = stm32_f1_pinval_get(pin);
+
+	u32_t temp = conf & (STM32_MODE_INOUT_MASK << STM32_MODE_INOUT_SHIFT);
+	if (temp == STM32_MODE_INPUT) {
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Break a line before "if..." for readability.

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+int stm32_gpio_configure(u32_t *base_addr, int pin, int conf, int altf)
+{
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)base_addr;
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(altf);
+
+	pin = stm32_f1_pinval_get(pin);
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Put this function common to all socs.

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+#else
+#define STM32_PINCFG_MODE_OUTPUT	STM32_MODER_OUTPUT_MODE
+#define STM32_PINCFG_MODE_INPUT		STM32_MODER_INPUT_MODE
+#define STM32_PINCFG_PULL_UP		STM32_PUPDR_PULL_UP
+#define STM32_PINCFG_PULL_DOWN		STM32_PUPDR_PULL_DOWN
+#define STM32_PINCFG_FLOATING		STM32_PUPDR_NO_PULL
+#endif
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+static inline u32_t stm32_f1_pinval_get(int pin)
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Get this function common to all SoCs.
Put  #îfdef internal to the function

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+			if (temp == STM32_PUPD_PULL_UP) {
+				LL_GPIO_SetPinPull(gpio, pin, LL_GPIO_PULL_UP);
+			} else {
+				LL_GPIO_SetPinPull(gpio, pin, LL_GPIO_PULL_DOWN);
+			}
+		}
+
+	} else {
+		temp = conf & (STM32_CNF_OUT_1_MASK << STM32_CNF_OUT_1_SHIFT);
+		if (temp == STM32_CNF_GP_OUTPUT) {
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Break a line be for if

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+			LL_GPIO_SetPinSpeed(gpio, pin, LL_GPIO_SPEED_FREQ_MEDIUM);
+		} else {
+			LL_GPIO_SetPinSpeed(gpio, pin, LL_GPIO_SPEED_FREQ_HIGH);
+		}
+	}
+#else
+	pin = 1 << pin;
+
+	u32_t temp = conf & (STM32_MODER_MASK << STM32_MODER_SHIFT);
+	if (temp == STM32_MODER_OUTPUT_MODE) {
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Break a line before if

File: soc/arm/st_stm32/common/soc_gpio.c [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+
+int stm32_gpio_configure(u32_t *base_addr, int pin, int conf, int altf)
+{
+	GPIO_TypeDef *gpio = (GPIO_TypeDef *)base_addr;
+
+#ifdef CONFIG_SOC_SERIES_STM32F1X
+	ARG_UNUSED(altf);
+
+	pin = stm32_f1_pinval_get(pin);
+
```
Comment (erwango) [inline soc/arm/st_stm32/common/soc_gpio.c]: Add some comments to this function.

**Review by erwango — COMMENTED** [review]
Here are some clues to make code more readable.

**Comment by dbkinder** [comment]
OK, I removed me as a reviewer.

**Comment by cybertale** [comment]
Hi @erwango ,
Could you offer some debug information if you got some time?
Here are my debug procedure. I believe you understand this better than me, I'm just ensure that we are using the same commands and see the same stuff.

In build directory of samples/basic/disco, after run 'ninja' and 'ninja flash'
run 
On F1 platform, run
`ninja debug`
`b soc_gpio.c:214`
`p gpio`
`p gpio->CRH`
`p gpio->CRL`
`c`
`p gpio`
`p gpio->CRH`
`p gpio->CRL`
On F7 platform, run
`ninja debug`
`b soc_gpio.c:214`
`p gpio`
`p gpio->MODER`
`c`
`p gpio`
`p gpio->MODER`

And of course the LEDs ports and pin numbers of your boards or what your boards are and I can look it up myself. 
Running into breakpoints twice because LEDs of some boards are in different port.
Though I don't have a F7 board(actually I have one, just got it lost in the lab...), I think the F4 board I have should act just the same.

**Comment by erwango** [comment]
@cybertale , won't have time to do it today sorry. Will have a try tomorrow

**Comment by cybertale** [comment]
No hurry.

**Comment by erwango** [comment]
Ok, so actually, to work with sample/disco on all boards requires a rebase...
And this is working!

Let me have a look to sample/basic/button

**Comment by erwango** [comment]
@psidhu , btw, you might be interested by that one.

**Comment by cybertale** [comment]
@erwango Great the disco sample works! My boards are from a Chinese manufacture and there aren't many information from Google. They are Atomic Explorer(F4) and Atomic Warship(F1).
Sorry there seems like is a bug in reading. I got it worked earlier and only tested the disco sample in the latest commit, I'm debugging it now.

**Comment by cybertale** [comment]
@erwango Sorry for my mistake, it was because a error in index translation in soc_gpio:252. I've fixed it, give the new code a try if you got some time!

File: soc/arm/st_stm32/stm32f0/soc.h [hunk 15]
```diff
@@ -55,6 +55,10 @@
 #include <stm32f0xx_ll_spi.h>
 #endif
 
+#ifdef CONFIG_GPIO_STM32
```
Comment (b0661) [inline soc/arm/st_stm32/stm32f0/soc.h]: Why not use CONFIG_USE_STM32_LL_GPIO instead?

It was created to flag the usage of LL_GPIO. Using CONFIG_GPIO_STM32 restricts the inclusion to the STM32_GPIO driver - but other drivers may also want to use the LL_GPIO functions.

Comment (cybertale) [reply soc/arm/st_stm32/stm32f0/soc.h]: Understand, while I think it's because for other drivers, the interface of enabling gpio driver of stm32 should be always CONFIG_GPIO_STM32. Other drivers like spi drivers should use generic configuration functions to configure the port, pin, and the gpio driver will handle the configuration. No matter if the GPIO driver is using the LL library.
There is another reason for it, other drivers define CONFIG_SPI_STM32, CONFIG_I2C_STM32_V2... 
So, I'm new to zephyr and just expressing what I'm thinking about this problem, and I think this may should be discussed in a new thread.

File: soc/arm/st_stm32/common/soc_gpio.c (hunk 3 above)
Comment (cybertale) [reply soc/arm/st_stm32/common/soc_gpio.c]: Yes, and should syscfg_registers.h be removed?? Since we prefer using the LL library so they may will not be used in the future.

File: soc/arm/st_stm32/common/soc_gpio.c (hunk 5 above)
Comment (cybertale) [reply soc/arm/st_stm32/common/soc_gpio.c]: So I'm thinking all functions in soc_gpio.c file can be moved into gpio_stm32.c file. Should moving them into gpio_stm32.c be a separate commit ?

File: soc/arm/st_stm32/stm32f0/soc.h (hunk 15 above)
Comment (erwango) [reply soc/arm/st_stm32/stm32f0/soc.h]: @b0661 , USE_STM32_LL_XXX is made for LL sources, not headers.
Using these flags in soc.h would work but it's a new way of using them and this deserve a dedicated discussion.

**Comment by erwango** [comment]
@cybertale , thanks for the change.
I've made few tests with samples/basic/button 
* Working on F1 / F4 / F7 / L4
* Not working on F0 / L0 / F3. On F3 UART is not working anymore

Seems that some bug(s?) remain. Can you have a check ?

**Comment by cybertale** [comment]
@erwango Strange, I checked with the reference manuals, F0/L0/F3 and F4/F7 have almost identical GPIO and SYSCFG->EXTICR[1-4] architecture. Only in F0/L0 series alternate functions are limited to 8 kind, and some initial values of MODER registers are different. 
For UART they are also the same. Currently I do not have any boards with F0/L0/F3 series of chips. 
Could you have a little debug on them if you got some time?
`ninja debug`
`b soc_gpio.c:106`
`p gpio`
`p gpio->MODER`
`p gpio->PUPDR`
print the layout of EXTICRs
`x 0x40010008`
`x 0x4001000c`
`x 0x40010010`
`x 0x40010014`

**Comment by erwango** [comment]
@cybertale 

Debug on nucleo_l073rz.
```
Breakpoint 1, stm32_gpio_configure (base_addr=base_addr@entry=0x50000000, pin=pin@entry=2, conf=conf@entry=548, altf=altf@entry=4) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:149
149		pin = 1 << pin;
(gdb) p gpio
$1 = (GPIO_TypeDef *) 0x50000000
(gdb) p pin
$2 = 2
(gdb) p /x (GPIO_TypeDef)*gpio
$3 = {MODER = 0xebfffcff, OTYPER = 0x0, OSPEEDR = 0xc000000, PUPDR = 0x24000000, IDR = 0x4000, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
(gdb) x 0x40010008
0x40010008:	0x00000000
(gdb) x 0x4001000C
0x4001000c:	0x00000000
(gdb) x 0x40010010
0x40010010:	0x00000000
(gdb) x 0x40010014
0x40010014:	0x00000000

```
I break run to see EXTI CR after configuration phase and still 0..
```
(gdb) x 0x40010008
0x40010008:	0x00000000
(gdb) x 0x4001000C
0x4001000c:	0x00000000
(gdb) x 0x40010010
0x40010010:	0x00000000
(gdb) x 0x40010014
0x40010014:	0x00000000
```

**Comment by cybertale** [comment]
@erwango  Odd, nucleo_l073rz's button is connected with pin 13 of gpioc, which base address should be 0x50000800, while result of your gpio is 0x50000000, which is the base address of GPIOA.

**Comment by erwango** [comment]
> @erwango Odd, nucleo_l073rz's button is connected with pin 13 of gpioc, which base address should be 0x50000800, while result of your gpio is 0x50000000, which is the base address of GPIOA.

Button samples enable UART So this is PA2 configuration... Need to test again

**Comment by erwango** [comment]
Beginning of function
```
Breakpoint 1, stm32_gpio_configure (base_addr=0x50000800, pin=pin@entry=13, conf=0, altf=altf@entry=0) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:149
149		pin = 1 << pin;
(gdb) p gpio
$1 = (GPIO_TypeDef *) 0x50000800
(gdb) p pin
$2 = 13
(gdb) p /x (GPIO_TypeDef)*gpio
$3 = {MODER = 0xffffffff, OTYPER = 0x0, OSPEEDR = 0x0, PUPDR = 0x0, IDR = 0x0, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
```
End of function:
```
Breakpoint 2, stm32_gpio_configure (base_addr=0x50000800, pin=8192, pin@entry=13, conf=0, altf=altf@entry=0) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:202
202		return 0;
(gdb) p /x (GPIO_TypeDef)*gpio
$4 = {MODER = 0xf3ffffff, OTYPER = 0x0, OSPEEDR = 0x0, PUPDR = 0x0, IDR = 0x2000, ODR = 0x0, BSRR = 0x0, LCKR = 0x0, AFR = {0x0, 0x0}, BRR = 0x0}
```

**Comment by cybertale** [comment]
@erwango Good! GPIO configuration seems fine, what about the EXTICRs?

**Comment by cybertale** [comment]
I'm not sure when the interrupt is enabled, I usually break at the reads of the port, interrupt should be initialized that time.

**Comment by erwango** [comment]
> I'm not sure when the interrupt is enabled, I usually break at the reads of the port, interrupt should be initialized that time.

I break after init. For F4 I can see config is ok. For L0, everything is 00

**Comment by erwango** [comment]
Btw, I'll be off for almost 2 weeks in few hours. I won't be able to help anymore during that time.

**Comment by cybertale** [comment]
OK, then I'd better take a good use of these hours! :)
Let me have a look at the code.

**Comment by cybertale** [comment]
Then it should be the problem of EXTI, could you help break at soc_gpio.c:258 and print port, pin, line? I think this is where problems should be.

**Comment by erwango** [comment]
```
Breakpoint 1, stm32_gpio_enable_int (port=2, pin=pin@entry=13) at /local/mcu/zephyr/zephyr-project/soc/arm/st_stm32/common/soc_gpio.c:260
260		if (port == STM32_PORTH) {
(gdb) p pin
$1 = 13
(gdb) p port
$2 = 2
(gdb) p /x line
$4 = 0xf00003

```

**Comment by cybertale** [comment]
Sorry, I didn't notice that there are difference in LL for l0 and f4, I found the problem.

**Comment by cybertale** [comment]
@erwango I fixed the problem, please give it a try if you got some time.

**Comment by cybertale** [comment]
I just change soc_gpio.c:243 from 
`line = (0xF << ((pin % 4 * 4) + 16)) | (pin / 4);`
to
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32F0X)
	line = ((pin % 4 * 4) << 16) | (pin / 4);
#else
	line = (0xF << ((pin % 4 * 4) + 16)) | (pin / 4);
#endif
```

**Comment by erwango** [comment]
Still KO

**Comment by cybertale** [comment]
Could you break at LL_SYSCFG_SetEXTISource(port, line) in stm32_gpio_enable_int and print out the pin and line?

**Comment by erwango** [comment]
Working when using the lines you provided, I had an issue when collecting the update change. My bad

**Comment by cybertale** [comment]
@erwango Great! My first experience  of remote debugging, happy it's working now. I'm thinking of merging all functions in soc_gpio.c into gpio_stm32.c, what do you think? I'm planning clean up the code as you mentioned now since you'll soon be off.

**Comment by erwango** [comment]
Great ! 
One remaining issue: UART Ko on STM32F3....

>  I'm thinking of merging all functions in soc_gpio.c into gpio_stm32.c

This makes perfect sense

**Comment by cybertale** [comment]
I'm thinking breaking after init, and print out the layout of gpio like you used:
`(gdb)` p /x (GPIO_TypeDef)*gpio`
and EXTICRs.
For F3, EXTICRs are still 
 0x40010008
 0x4001000c
 0x40010010
 0x40010014

**Comment by erwango** [comment]
> For F3, EXTICRs are still

I've did a quick test and EXTI are fine for F3, blinky as well. Only issue it UART.

**Comment by erwango** [comment]
Ok, strange.
This is working with nucleo_f334r8/nucleo_f302r8.
But on stm32f3_disco uart is not working with your change (while ok on master).

**Comment by cybertale** [comment]
Is EXTI fine on stm32f3_disco?

**Comment by cybertale** [comment]
UART problem should be related to afrl and afrh...I think

**Comment by erwango** [comment]
> Is EXTI fine on stm32f3_disco?

Yes, I tested with modified sample, and it's working.
Should be UART pin config issue

**Comment by cybertale** [comment]
It's better to see the full gpio registers layout of stm32f3_disco.

**Comment by cybertale** [comment]
I mean after init, what afr[0]/[1] and moder of it contains.

**Comment by erwango** [comment]
So console is UART1, using GPIOC pins:
```
(gdb) p /x *0x48000800  (MODER)
$2 = 0xa00000
(gdb) p /x *0x48000820 (AFRL)
$3 = 0x0
(gdb) p /x *0x48000824 (AFRH)
$4 = 0x7700
```

**Comment by cybertale** [comment]
So AFRH should be end with 770 not 7700 if I'm calculating this right.
pin 9/10 should be using AF 7, right?

**Comment by erwango** [comment]
> pin 9/10 should be using AF 7, right?

Exact

**Comment by erwango** [comment]
Wait: it's actually:
```
	{STM32_PIN_PC4, STM32F3_PINMUX_FUNC_PC4_USART1_TX},
	{STM32_PIN_PC5, STM32F3_PINMUX_FUNC_PC5_USART1_RX},
```

**Comment by erwango** [comment]
Issue is in stm32_gpio_configure at following line 
```
		if (pin < 8) {
```
pin is not pin anymore it's 1 << pin.

**Comment by erwango** [comment]
tested ok with  a quick fix.
I'm going now!

**Comment by cybertale** [comment]
Oh, sorry it's my fault! Great it's working!

**Comment by cybertale** [comment]
Have a good trip! I'll try to clean up the code and manage to take care the thing you mentioned in another thread while you're away.

File: drivers/gpio/Kconfig.stm32 [hunk 16]
```diff
@@ -9,6 +9,7 @@ menuconfig GPIO_STM32
 	bool "GPIO Driver for STM32 family of MCUs"
 	depends on GPIO && SOC_FAMILY_STM32
 	select HAS_DTS_GPIO
+	select USE_STM32_LL_GPIO
```
Comment (erwango) [inline drivers/gpio/Kconfig.stm32]: You don't actually need that for now.
This is made to compile the ll .c files which are not used in the drivers for now (only functions defined in ll header files are used).

**Review by erwango — CHANGES_REQUESTED** [review]
Tested ok on all stm32 series available in Zephyr today.
One small fix and we're good!
Thanks again for the change, this saves a lot of code to maintain and even saves few cycles at execution.

**Comment by cybertale** [comment]
Hi @erwango ,I removed it, how about this time?

**Comment by cybertale** [comment]
And I didn't quit understand what should I do with #12177. Power operations are additions above my code, so there's nothing I should do, right?

**Comment by erwango** [comment]
Ok, it just depends which one gets merged first. Let me ack this one and we'll see

**Comment by erwango** [comment]
ok, seems a rebase is required first

**Comment by erwango** [comment]
@cybertale , required rebase is due to commit https://github.com/zephyrproject-rtos/zephyr/commit/359adfc9598758c5d35965473abd897813e709e2 from https://github.com/zephyrproject-rtos/zephyr/pull/11667 which is applying MISRA-C rule about boolean operands evaluation. Since you're deleting the files, it doesn't matter much but please take these rule into account in the code you're adding.

**Comment by cybertale** [comment]
Hi @erwango, I updated the code, please give it a look if you got some time!

File: drivers/gpio/gpio_stm32.c [hunk 17]
```diff
@@ -37,6 +36,226 @@ static void gpio_stm32_isr(int line, void *arg)
...
+			LL_GPIO_SetPinPull(gpio, pin, LL_GPIO_PULL_DOWN);
+		} else {
+			LL_GPIO_SetPinPull(gpio, pin, LL_GPIO_PULL_NO);
+		}
+	} else if (temp == STM32_MODER_ANALOG_MODE) {
+		LL_GPIO_SetPinMode(gpio, pin, LL_GPIO_MODE_ANALOG);
+	} else {
+		LL_GPIO_SetPinMode(gpio, pin, LL_GPIO_MODE_ALTERNATE);
+
+		if (pin < (1 << 8)) {
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c]: This is correct but confusing. 
We should introduce an alternate variable for pin used with LL api  (pin_ll ? Anything better is welcome) so we could use a simple and clear if statement here (if pin < 8).

File: drivers/gpio/gpio_stm32.c [hunk 18]
```diff
@@ -54,12 +273,12 @@ static int gpio_stm32_config(struct device *dev, int access_op,
...
 	 * configuration
 	 */
-	map_res = stm32_gpio_flags_to_conf(flags, &pincfg);
+	map_res = gpio_stm32_flags_to_conf(flags, &pincfg);
 	if (map_res) {
 		return map_res;
 	}
 
-	if (stm32_gpio_configure(cfg->base, pin, pincfg, 0)) {
+	if (gpio_stm32_configure(cfg->base, pin, pincfg, 0)) {
```
Comment (erwango) [inline drivers/gpio/gpio_stm32.c]: Add a "!= 0" to the if statement

**Review by erwango — CHANGES_REQUESTED** [review]
Last comments and we should be good.

**Comment by cybertale** [comment]
Hi @erwango , I added some '!=0' and '== 0' to some if in gpio_stm32.c and the additional pin_ll variable, how about this time?

**Review by erwango — APPROVED** [review]
Tested on 10 boards (at least one from each supported STM32 series)
